keps/sig-scheduling/3370-native-podgroup-api/README.md

```diff
@@ -0,0 +1,1076 @@
+  - Compare their pod group's `CreationTimestamp`: the Pod in pod group which has earlier timestamp is positioned ahead of the other.
+  - If their `CreationTimestamp` is identical, order by their UID of PodGroup: a Pod with lexicographically greater UID is scheduled ahead of the other Pod. (The purpose is to tease different PodGroups with the same `CreationTimestamp` apart, while also keeping Pods belonging to the same PodGroup back-to-back)
+
+#### PreFilter
+In PreFilter phase, we need to check if pod.spec.podGroup is nil firstly. If so, return success. If not, check if the pod group is existed in the scheduler internal cache and pod.spec.podGroup.subset is matched in pod group, if not, reject the pod in PreFilter.
+
+
+#### Permit
+In Permit phase, It is divided into two parts.
+
+- The first part is that we put the pods that doesn't meet MinMember into the WaitingMap and reserve resources until MinMember are met or timeout is triggered.
+
+   - Get the number of Running pods that belong to the same PodGroup
+   - Get the number of WaitingPods (used to record pods in waiting status) that belong to the same PodGroup
+
+  If pods in all the subsets meet `Running + WaitingPods + 1 >= MinMember(1 means the pod itself) `, approve the waiting pods that belong to the same PodGroup. Otherwise, put the pod into WaitingPods and set the timeout (the timeout comes from the ScheduleTimeoutSeconds in PodGroup.).
+
+- Then second part is in order for pods belonging to the same PodGroup to be scheduled as a whole without being broken up, we will leverage the feature added https://github.com/kubernetes/kubernetes/pull/103383 to move pods of the same PodGroup back to internal scheduling activeQ instantly.
+
+#### UnReserve
+After a pod which belongs to a PodGroup times out in the permit phase. UnReserve Rejects the pods that belong to the same PodGroup to avoid long-term invalid reservation of resources. The rejected pod will have the different failure message in condition with other scheduled failed pods.
+
+```go
+  // These are reasons for a pod's transition to a condition.
+  const (
+    // PodReasonUnschedulable reason in PodScheduled PodCondition means that the scheduler
+    // can't schedule the pod right now, for example due to insufficient resources in the cluster.
+    PodReasonUnschedulable = "Unschedulable"
+
+    // PodReasonCoschedulingNotMeetMinMember reason in PodScheduled PodCondition means 
+    // that the pod can be scheduled successfully, but the scheduling fails because 
+    // minMember cannot be met.
+    PodReasonCoschedulingNotMeetMinMember = "CoschedulingNotMeetMinMember"
+  )
+
+
+  pod.status.conditions[*].reason = PodReasonCoschedulingNotMeetMinMember
+```
+### Cluster Autoscaler
+
+Cluster Autoscaler check the failure reasons in pod.status.condition to decide whether it's possible to resolve the failure by provisioning new machines. If the reason is `CoschedulingNotMeetMinMember`, CA don't need to provision new machines.
```

Comment:
It maybe like this in autoscaler: https://github.com/kubernetes/autoscaler/blob/6558beddd3d50c4653f45bad7958bf41e050f005/cluster-autoscaler/core/static_autoscaler.go#L405
We can filter the pods with `CoschedulingNotMeetMinMember` in autoscaler.

Reply:
Wouldn't filtering these pods cause them to wait for scale up indefinitely? CA should provision nodes to schedule the entire PodGroup if possible, but it quite unclear to me if it should provision partial resources: in some scenarios that might be desirable, e.g. when different Subsets require scaling up different NodeGroups (in which case multiple scale ups are required).

Reply:
A fool question, does CA support scaling up with specific NodePools now? As you said, I have particular resources requirements. @x13n

Reply:
NodePool is not a k8s concept. That being said, CA operates on NodeGroups, since most cloud providers implement such concept (e.g. nodepools in GKE). It is quite possible for nodes in one NodeGroup to have certain label specified, so let's say:
- `pod1` has nodeSelector `label1=value1`
- `pod2` has nodeSelector `label2=value2`
- nodes in `nodeGroup1` have `label1=value1`
- nodes in `nodeGroup2` have `label2=value2`

In each iteration, CA picks the best node group to scale up. In the scenario above, neither `nodeGroup1` nor `nodeGroup2` can satisfy both `pod1` and `pod2`, so 2 subsequent scale ups are required. If these pods belong to the same PodGroup, the resources for them need to be provisioned incrementally. Subsequent scheduling will follow all-or-nothing semantics, so it will wait for all scale ups to take place.